### PR TITLE
Revise Azure CI scripts

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -1,130 +1,73 @@
...
-      vmImage: 'vs2017-win2016'
-    variables:
-      VSINSTALLDIR: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\
-      HOST_DMD_VERSION: 2.090.0
-    strategy:
-      matrix:
-        win32-ldc:
+        x86-mscoff:
+          OS: Win_32
+          MODEL: 32mscoff
```
Comment (kinke) [inline azure-pipelines.yml:65]: The `MODEL` for this job used to be `32` - but was overridden to `32mscoff` in https://github.com/dlang/dmd/blob/a0875a7e0bb9cd27f12b09eb96c391200a743034/.azure-pipelines/windows-msbuild.bat#L8-L11

**Comment by kinke** [comment]
Looks like the *required* CI jobs need an adaptation after renaming the Azure jobs.

**Comment by thewilsonator** [comment]
LMK when this is ready for review.

**Comment by kinke** [comment]
This should be it for the moment. - Getting the Phobos unittests to compile for 32-bit OMF would require adapting the Phobos makefile, splitting it up into more object files.

File: .azure-pipelines/windows.sh [hunk 2]
```diff
@@ -75,73 +71,82 @@ else
...
-# REASON: whitespace path variables in DigitalMars make from indirect invocation from Phobos
-################################################################################
-
-if [ "$MODEL" != "32" ] ; then
-    cd "${DMD_DIR}/../phobos/etc/c/zlib"
-    "${DM_MAKE}" -f win64.mak MODEL=${MODEL} "zlib${MODEL}.lib" "CC=$CC" "LIB=$AR" VCDIR=. MAKE="$DM_MAKE"
-fi
+cd "$DMD_DIR/src"
+"$DM_MAKE" -f "$MAKE_FILE" MAKE="$DM_MAKE" BUILD=debug unittest
+DFLAGS="-L-LARGEADDRESSAWARE" "$DM_MAKE" -f "$MAKE_FILE" MAKE="$DM_MAKE" reldmd-asserts
```
Comment (kinke) [inline .azure-pipelines/windows.sh:81]: `/LARGEADDRESSAWARE` was needed to overcome process memory limits when compiling the Phobos unittests with a 32-bit DMD (before hitting the max 32K symbols in OMF object file limitation). The leading dash somehow led to something prepending a path (=> something like `/C/Program Files/Git/LARGEADDRESSAWARE`), hence the `-`, which fortunately works for both optlink.exe and link.exe.

File: src/win32.mak [hunk 3]
```diff
@@ -125,8 +125,8 @@ profile:
 
 trace: debdmd-make
 
-unittest:
-	$(DMDMAKE) "ENABLE_DEBUG=1" "ENABLE_UNITTEST=1" "ENABLE_COVERAGE=1" $(TARGETEXE)
+unittest: $(GEN)\build.exe
+	$(RUN_BUILD) unittest
```
Comment (kinke) [inline src/win32.mak:129]: This is analogous to posix.mak; the previous target built a normal DMD with enabled unittests (and building the default target afterwards was a no-op...).

File: test/run.d [hunk 4]
```diff
@@ -143,6 +143,13 @@ Options:
         return spawnProcess(unitTestRunnerCommand ~ args).wait();
     }
 
+    if (args == ["tools"])
+    {
+        verifyCompilerExists(env);
+        ensureToolsExists(env, EnumMembers!TestTools);
+        return 0;
+    }
```
Comment (kinke) [inline test/run.d:151]: This allows to prebuild the test tools via `run tools`, which simplifies some scenarios, where e.g. environment variables are needed for the DMD/optlink/... to be tested, but would also affect the host compiler.

File: test/tools/d_do_test.d [hunk 5]
```diff
@@ -803,20 +803,13 @@ bool collectExtraSources (in string input_dir, in string output_dir, in string[]
...
-            }
-            else if (envData.os == "windows" && envData.model == "32")
-            {
-                command ~= " -c "~curSrc~" -o"~curObj;
-            }
-            else
-            {
-                command ~= " -m"~envData.model~" -c "~curSrc~" -o "~curObj;
-            }
+            command ~= ` /c /nologo `~curSrc~` /Fo`~curObj;
```
Comment (kinke) [inline test/tools/d_do_test.d:808]: E.g., LDC needs this too.

File: .azure-pipelines/windows-msbuild.bat [hunk 6]
```diff
@@ -16,21 +16,17 @@ set N=3
 set DM_MAKE=%DMD_DIR%\dm\path\make.exe
 set LDC_DIR=%DMD_DIR%\ldc2-%LDC_VERSION%-windows-multilib
 
+if "%D_COMPILER%" == "ldc" set HOST_DMD=%LDC_DIR%\bin\ldmd2.exe
+if "%D_COMPILER%" == "dmd" set HOST_DMD=%DMD_DIR%\dmd2\windows\bin\dmd.exe
+
 REM take the first found cl.exe, in case there was already one in the path when vcvarsall.bat was called
-FOR /F "tokens=* USEBACKQ" %%F IN (`where cl.exe`) DO (SET MSVC_CC=%%~fsF && goto CC_DONE)
+FOR /F "tokens=* USEBACKQ" %%F IN (`where cl.exe`) DO (SET MSVC_CC=%%~fsF
+  goto CC_DONE)
```
Comment (kinke) [inline .azure-pipelines/windows-msbuild.bat:24]: Previously, `MSVC_CC` included a trailing space, which broke the `usingMicrosoftCompiler` detection in `run.d`...

**Comment by kinke** [comment]
Ping. Among other things, this makes sure Phobos is actually CI-tested on Win64 - the phobos repo currently isn't (!).

**Comment by rainers** [comment]
LGTM, but I only made a cursory pass. There are duplicate Azure pipeline checks AFAICT, though. Or is this just a temporary hickup?

**Comment by kinke** [comment]
They aren't duplicates; the 'required' CI jobs need to be renamed accordingly (no idea where though), as stated in https://github.com/dlang/dmd/pull/11795#issuecomment-698978308.

**Comment by kinke** [comment]
> so that regressions would show up for the dmd repo alone

Looks like there has been a Phobos regression on Win64 in the meantime, because currently the `Windows_LDC_Debug x64-debug-ldc` job is the only one testing Phobos on Win64, and that one's failing for master.

**Review by Geod24 — APPROVED** [review]
Thanks for taking care of this. I have changed the branch requirements so a rename won't be a problem in the future (as long as we can rely on "Azure Pipelines" being an umbrella).

**Pull request merged**